Realm/RLMLogger.mm

```diff
@@ -0,0 +1,171 @@
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMLogger_Private.h"
+
+#import "RLMUtil.hpp"
+
+#import <realm/util/logger.hpp>
+
+typedef void (^RLMLoggerFunction)(RLMLogLevel level, NSString *message);
+
+using namespace realm;
+using Logger = realm::util::Logger;
+using Level = Logger::Level;
+
+namespace {
+static Level levelForLogLevel(RLMLogLevel logLevel) {
+    switch (logLevel) {
+        case RLMLogLevelOff:    return Level::off;
+        case RLMLogLevelFatal:  return Level::fatal;
+        case RLMLogLevelError:  return Level::error;
+        case RLMLogLevelWarn:   return Level::warn;
+        case RLMLogLevelInfo:   return Level::info;
+        case RLMLogLevelDetail: return Level::detail;
+        case RLMLogLevelDebug:  return Level::debug;
+        case RLMLogLevelTrace:  return Level::trace;
+        case RLMLogLevelAll:    return Level::all;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static RLMLogLevel logLevelForLevel(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:    return RLMLogLevelOff;
+        case Level::fatal:  return RLMLogLevelFatal;
+        case Level::error:  return RLMLogLevelError;
+        case Level::warn:   return RLMLogLevelWarn;
+        case Level::info:   return RLMLogLevelInfo;
+        case Level::detail: return RLMLogLevelDetail;
+        case Level::debug:  return RLMLogLevelDebug;
+        case Level::trace:  return RLMLogLevelTrace;
+        case Level::all:    return RLMLogLevelAll;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static NSString* levelPrefix(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:
+        case Level::all:    return @"";
+        case Level::trace:  return @"Trace";
+        case Level::debug:  return @"Debug";
+        case Level::detail: return @"Detail";
+        case Level::info:   return @"Info";
+        case Level::error:  return @"Error";
+        case Level::warn:   return @"Warning";
+        case Level::fatal:  return @"Fatal";
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+struct CocoaLogger : public Logger {
+    void do_log(Level level, const std::string& message) override {
+        NSLog(@"%@: %@", levelPrefix(level), RLMStringDataToNSString(message));
+    }
+};
+
+class CustomLogger : public Logger {
+public:
+    RLMLoggerFunction function;
+    void do_log(Level level, const std::string& message) override {
+        @autoreleasepool {
+            if (function) {
+                function(logLevelForLevel(level), RLMStringDataToNSString(message));
+            }
+        }
+    }
+};
+} // anonymous namespace
+
+@implementation RLMLogger {
+    std::shared_ptr<Logger> _logger;
+}
+
+static RLMLogger *s_defaultLogger;
+
+typedef void(^LoggerBlock)(RLMLogLevel level, NSString *message);
+
+- (RLMLogLevel)level {
+    return logLevelForLevel(Logger::get_default_level_threshold());
```

Comment:
Now that these are instance methods they should be getting and setting the level on the instance's logger.